Add an "Implementing a Compatible Data Plane" section to the implementers guide (#1143)
kubernetes-sigs/gateway-api-inference-extension: site-src/guides/implementers.md

  extensionRef:
    name: vllm-llama3-8b-instruct-epp
```

> mkdocs.yml

Reviewer: Guessing this was unintentional? Since we already split conformance tests out into a separate guide, we could do the same here and make a corresponding change to mkdocs.yml. Don't think this is a blocker, will defer to you here.

The EPP communicates the chosen endpoint to the proxy via the `x-gateway-destination-endpoint` HTTP header and the `dynamic_metadata` field of the ext-proc response. Failure to communicate the endpoint using both methods results in a 503 error if no endpoints are ready, or a 429 error if the request should be dropped. The header and metadata values must match. In addition to the chosen endpoint, a single fallback endpoint CAN be set using the key `x-gateway-destination-endpoint-fallback` in the same metadata namespace as one used for `x-gateway-destination-endpoint`.

### Implementing a Compatible Data Plane

To conform with the Inference Extensions API, Gateway data planes must implement the [Endpoint Picker Protocol](https://github.com/kubernetes-sigs/gateway-api-inference-extension/tree/main/docs/proposals/004-endpoint-picker-protocol).

At a high level, the protocol consists of metadata key/value pairs exchanged between the data plane and extensions containing relevant endpoint selection information:

> - From extension to data plane: the metadata contains the selected endpoints.
> - From data plane to extension: the metadata contains an optional subset of endpoints that the extension should pick from.

Reviewer: Unfortunately mkdocs list formatting is pretty unforgiving. Generally need an empty line before any list starts: https://deploy-preview-1143--gateway-api-inference-extension.netlify.app/guides/implementers/#implementing-a-compatible-data-plane. Same comment applies throughout, will likely take some trial and error.
```suggestion

- From extension to data plane: the metadata contains the selected endpoints.
- From data plane to extension: the metadata contains an optional subset of endpoints that the extension should pick from.
```

The key requirements for implementing the GIE protocol are as follows:
- Relies on the [ext_proc (External Processing)](https://www.envoyproxy.io/docs/envoy/latest/configuration/http/http_filters/ext_proc_filter) protocol as the foundation for exchanging HTTP stream payload and metadata throughout the various HTTP lifecycle events; several key details:
  - ext_proc relies on gRPC (bidirectional streaming) as the transport protocol

>   - ext_proc supports several processing modes, including buffered and streaming options for payload exchange

Reviewer 2: Should probably be mentioned that GIE only support FULL_DUPLEX_STREAMED now, full duplex uses specific proto objects so there isn't cross compatibility with the less strict options. 

I think that could be changed upstream however.

Author: Discussed with Kellen offline -- the requirement for FULL_DUPLEX_STREAMED is not well specified right now. Let's deal with this in the EPP protocol proposal itself.

  - Supports generic “service callouts” not restricted to genAI serving or AI use cases; e.g., mutation of cache keys for caching.
- [Alibaba Cloud](https://www.alibabacloud.com/help/en/asm/user-guide/use-envoy-external-processing-for-custom-processing-of-requests)
- GenAI serving
    - [AIBrix](https://aibrix.readthedocs.io/latest/features/gateway-plugins.html)
      - Enables inference optimized routing for the Gateway in Bytedance’s genAI inference infrastructure.
    - [Envoy AI Gateway](https://aigateway.envoyproxy.io/docs/concepts/architecture/data-plane)
      - Enables AI model based routing, request transformations and upstream authn.
- [Atlassian Guard](https://www.atlassian.com/software/guard)

Supporting this broad range of extension capabilities (including for inference, as evidenced above) requires hooks into all HTTP stream (i.e., request and response) lifecycle events as well as the corresponding headers, trailers and payload. This is the core value proposition for ext_proc, along with configurable options (such as for buffering and streaming modes) that enable its use across a variety of deployment scenarios and networking topologies.

> #### Native Implementations

Reviewer: ```suggestion
#### Open Source Implementations
```

> Several native implementations can be used as references:

Reviewer: ```suggestion
Several open source implementations can be used as references:
```
